pkg/retrieval/retrieval.go

```diff
@@ -90,32 +98,85 @@ func (s *Service) RetrieveChunk(ctx context.Context, addr swarm.Address) (swarm.
 		span, logger, ctx := s.tracer.StartSpanFromContext(ctx, "retrieve-chunk", s.logger, opentracing.Tag{Key: "address", Value: addr.String()})
 		defer span.Finish()
 
-		var skipPeers []swarm.Address
+		sps := newSkipPeersService()
 
-	LOOP:
-		for i := 0; i < maxPeers; i++ {
-			select {
-			case <-ctx.Done():
-				break LOOP
-			default:
+		ticker := time.NewTicker(retrieveRetryIntervalDuration)
+
+		var (
+			peerAttempt int
+			resultC     = make(chan retrieveChunkResult)
+			wgChan      sync.WaitGroup
+		)
+
+		wgChan.Add(maxPeers)
+
+		defer func() {
+			if peerAttempt != maxPeers {
```

Comment:
Since we're forwarding a request from another peer, we would like to return the chunk and forward it back as fast as possible. Blocking until all attempts finish should then be done in a separate goroutine. However, this might result in a bit of an ugly and convoluted goroutine management. It might benefit to switch to a channel driven approach here in order to communicate chunk or error, but it would infer we need to do a bit more indirection in this package in order to expose `RetrieveChunk` as synchronous. Please see [fetcher logic](https://github.com/ethersphere/swarm/blob/ae39aaa6dbdbe0df2278dfea645f143c8cd6f3a3/storage/netstore.go) in the old swarm codebase.

Reply:
Initial change I made in fact did returned as soon as chunk was retrieved from some peer.

Main problem in this case is the `context` that is being propagated to peer requests. Once chunk is returned the context is canceled, and "all" in-flight operations will be canceled.
I do not think that here some elaborate goroutines are required, just separation of context-es that are being used (which may or may not be a problem).